File: backend/src/tasks/es-client.ts
import { Client } from '@elastic/elasticsearch';
import { Domain, Organization } from '../models';

export const DOMAINS_INDEX = 'domains-5';
export const ORGANIZATIONS_INDEX = 'organizations-1';

interface Suggest {
  input: string | string[];
  weight: number;
}
interface DomainRecord extends Domain {
  suggest: Suggest[];
  parent_join: 'domain';
}

interface OrganizationRecord extends Organization {
  suggest: Suggest[];
}

export interface WebpageRecord {
  webpage_id: string;
  webpage_createdAt: Date | null;
  webpage_updatedAt: Date | null;
  webpage_syncedAt: Date | null;
  webpage_lastSeen: Date | null;
  webpage_url: string;
  webpage_status: string | number;
  webpage_domainId: string;
  webpage_discoveredById: string;
  webpage_responseSize: number | null;
  webpage_headers: { name: string; value: string }[];

  // Added before elasticsearch insertion (not present in the database):
  suggest?: { input: string | string[]; weight: number }[];
  parent_join?: {
    name: 'webpage';
    parent: string;
  };
  webpage_body?: string;
}

const organizationMapping = {
  name: {
    type: 'text'
  }
};

/**
 * Elasticsearch client.
 */
class ESClient {
  client: Client;

  constructor(isLocal?: boolean) {
    this.client = new Client({ node: process.env.ELASTICSEARCH_ENDPOINT! });
  }

  async syncOrganizationsIndex() {
    console.log('syncOrganizationsIndex');
    try {
      await this.client.indices.get({
        index: ORGANIZATIONS_INDEX
      });
      await this.client.indices.putMapping({
        index: ORGANIZATIONS_INDEX,
        body: { properties: organizationMapping }
      });
    } catch (e) {
      console.log('sync orgs error', e);
      await this.client.indices.create({
        index: ORGANIZATIONS_INDEX,
        body: {
          mappings: {
            properties: {
              ...organizationMapping,
              suggest: {
                type: 'completion'
              }
            },
            dynamic: true
          },
          settings: {
            number_of_shards: 2
          }
        }
      });
    }
  }

  /**
   * Creates the domains index, if it doesn't already exist.
   */
  async syncDomainsIndex() {
    console.log('syncDomainsIndex');
    const mapping = {
      services: {
        type: 'nested'
      },
      vulnerabilities: {
        type: 'nested'
      },
      webpage_body: {
        type: 'text',
        term_vector: 'yes'
      },
      parent_join: {
        type: 'join',
        relations: {
          domain: ['webpage']
        }
      }
    };
    try {
      await this.client.indices.get({
        index: DOMAINS_INDEX
      });
      await this.client.indices.putMapping({
        index: DOMAINS_INDEX,
        body: { properties: mapping }
      });
      console.log(`Index ${DOMAINS_INDEX} updated.`);
    } catch (e) {
      if (e.meta?.body?.error.type !== 'index_not_found_exception') {
        console.error(e.meta?.body);
        throw e;
      }
      await this.client.indices.create({
        index: DOMAINS_INDEX,
        body: {
          mappings: {
            properties: {
              ...mapping,
              suggest: {
                type: 'completion'
              }
            },
            dynamic: true
          },
          settings: {
            number_of_shards: 2
          }
        }
      });
      console.log(`Created index ${DOMAINS_INDEX}.`);
    }
    await this.client.indices.putSettings({
      index: DOMAINS_INDEX,
      body: {
        settings: { refresh_interval: '1800s' }
      }
    });
    console.log(`Updated settings for index ${DOMAINS_INDEX}.`);
  }

  excludeFields = (domain: Domain) => {
    const copy: any = domain;
    delete copy.ssl;
    delete copy.censysCertificatesResults;
    for (const service in copy.services) {
      delete copy.services[service].censysIpv4Results;
      delete copy.services[service].wappalyzerResults;
      delete copy.services[service].intrigueIdentResults;
    }
    return copy;
  };

  async updateOrganizations(organizations: Organization[]) {
    const organizationRecords = organizations.map((rec) => {
      return {
        suggest: [{ input: rec.name, weight: 1 }],
        ...rec
      };
    }) as OrganizationRecord[];
    const b = this.client.helpers.bulk<OrganizationRecord>({
      datasource: organizationRecords,
      onDocument(organization) {
        return [
          {
            update: {
              _index: ORGANIZATIONS_INDEX,
              _id: organization.id
            }
          },
          {
            doc_as_upsert: true
          }
        ];
      }
    });
    const result = await b;
    if (result.aborted) {
      console.error(result);
      throw new Error('Bulk operation aborted');
    }
    return result;
  }

  /**
   * Updates the given domains, upserting as necessary.
   * @param domains Domains to insert.
   */
  async updateDomains(domains: Domain[]) {
    console.log('updateDomains');
    const domainRecords = domains.map((e) => ({
      ...this.excludeFields(e),
      suggest: [{ input: e.name, weight: 1 }],
      parent_join: 'domain'
    })) as DomainRecord[];
    const b = this.client.helpers.bulk<DomainRecord>({
      datasource: domainRecords,
      onDocument(domain) {
        return [
          {
            update: {
              _index: DOMAINS_INDEX,
              _id: domain.id
            }
          },
          { doc_as_upsert: true }
        ];
      },
      onDrop(doc) {
        console.error(doc.error, doc.document);
        b.abort();
      }
    });
    const result = await b;
    if (result.aborted) {
      console.error(result);
      throw new Error('Bulk operation aborted');
    }
    return result;
  }

  /**
   * Updates the given webpages, upserting as necessary.
   * @param webpages Webpages to insert.
   */
  async updateWebpages(webpages: WebpageRecord[]) {
    console.log('updateWebpages');
    const webpageRecords = webpages.map((e) => ({
      ...e,
      suggest: [{ input: e.webpage_url, weight: 1 }],
      parent_join: {
        name: 'webpage',
        parent: e.webpage_domainId
      }
    })) as WebpageRecord[];
    const b = this.client.helpers.bulk<WebpageRecord>({
      datasource: webpageRecords,
      onDocument(webpage) {
        return [
          {
            update: {
              _index: DOMAINS_INDEX,
              _id: 'webpage_' + webpage.webpage_id,
              routing: webpage.webpage_domainId
            }
          },
          { doc_as_upsert: true }
        ];
      },
      onDrop(doc) {
        console.error(doc.error, doc.document);
        b.abort();
      }
    });
    const result = await b;
    if (result.aborted) {
      console.error(result);
      throw new Error('Bulk operation aborted');
    }
    return result;
  }

  /**
   * Searches for domains.
   * @param body Elasticsearch query body.
   */
  async searchDomains(body: any) {
    console.log('searchDomains');
    return this.client.search({
      index: DOMAINS_INDEX,
      body
    });
  }

  async searchOrganizations(body: any) {
    return this.client.search({
      index: ORGANIZATIONS_INDEX,
      body
    });
  }

  /**
   * Deletes everything in Elasticsearch
   */
  async deleteAll() {
    await this.client.indices.delete({
      index: '*'
    });
  }
}

export default ESClient;

File: backend/src/xfd_django/xfd_api/tasks/es_client.py
"""ES client."""
# Standard Python Libraries
import logging
import os
import time

# Third-Party Libraries
from elasticsearch import Elasticsearch, helpers
from elasticsearch.exceptions import TransportError

# Constants
DOMAINS_INDEX = "domains-5"
ORGANIZATIONS_INDEX = "organizations-1"

# Define mappings
organization_mapping = {
    "properties": {"name": {"type": "text"}, "suggest": {"type": "completion"}}
}

domain_mapping = {
    "properties": {
        "services": {"type": "nested"},
        "vulnerabilities": {"type": "nested"},
        "webpage_body": {"type": "text", "term_vector": "yes"},
        "parent_join": {"type": "join", "relations": {"domain": "webpage"}},
        "suggest": {"type": "completion"},
    }
}
LOGGER = logging.getLogger(__name__)
# Raise log level for Elasticsearch client to WARNING to suppress request logs
logging.getLogger("elasticsearch").setLevel(logging.WARNING)

# Also suppress low-level logs from urllib3 used by Elasticsearch
logging.getLogger("urllib3.connectionpool").setLevel(logging.WARNING)


class ESClient:
    """ES Client."""

    def __init__(self):
        """Initialize the Elasticsearch client."""
        endpoint = os.getenv("ELASTICSEARCH_ENDPOINT")
        self.client = Elasticsearch(endpoint)

    def sync_organizations_index(self):
        """Create or updates the organizations index with mappings."""
        try:
            if not self.client.indices.exists(index=ORGANIZATIONS_INDEX):
                LOGGER.info("Creating index %s...", ORGANIZATIONS_INDEX)
                self.client.indices.create(
                    index=ORGANIZATIONS_INDEX,
                    body={
                        "mappings": organization_mapping,
                        "settings": {"number_of_shards": 2},
                    },
                )
            else:
                LOGGER.info("Updating index %s...", ORGANIZATIONS_INDEX)
                self.client.indices.put_mapping(
                    index=ORGANIZATIONS_INDEX, body=organization_mapping
                )
        except Exception as e:
            LOGGER.error("Error syncing organizations index: %s", e)
            raise e

    def sync_domains_index(self):
        """Create or updates the domains index with mappings."""
        try:
            if not self.client.indices.exists(index=DOMAINS_INDEX):
                LOGGER.info("Creating index %s...", DOMAINS_INDEX)
                self.client.indices.create(
                    index=DOMAINS_INDEX,
                    body={
                        "mappings": domain_mapping,
                        "settings": {"number_of_shards": 2},
                    },
                )
            else:
                LOGGER.info("Updating index %s...", DOMAINS_INDEX)
                self.client.indices.put_mapping(
                    index=DOMAINS_INDEX, body=domain_mapping
                )
            # Set refresh interval
            self.client.indices.put_settings(
                index=DOMAINS_INDEX, body={"settings": {"refresh_interval": "1800s"}}
            )
        except Exception as e:
            LOGGER.error("Error syncing domains index: %s", e)
            raise e

    def update_organizations(self, organizations):
        """Update or inserts organizations into Elasticsearch."""
        actions = [
            {
                "_op_type": "update",
                "_index": ORGANIZATIONS_INDEX,
                "_id": org["id"],
                "doc": {**org, "suggest": [{"input": org["name"], "weight": 1}]},
                "doc_as_upsert": True,
            }
            for org in organizations
        ]
        self._bulk_update(actions)

    def update_domains(self, domains, max_retries=5, backoff_base=2):
        """Update or insert domains into Elasticsearch with retry and backoff."""
        actions = [
            {
                "_op_type": "update",
                "_index": DOMAINS_INDEX,
                "_id": domain["id"],
                "doc": {
                    **domain,
                    "suggest": [{"input": domain["name"], "weight": 1}],
                    "parent_join": "domain",
                },
                "doc_as_upsert": True,
            }
            for domain in domains
        ]

        attempt = 0
        while attempt <= max_retries:
            try:
                success, response = helpers.bulk(
                    self.client,
                    actions,
                    raise_on_error=False,
                    raise_on_exception=False,
                    request_timeout=60,
                )

                failed = [
                    item
                    for item in response
                    if "update" in item and item["update"].get("error")
                ]
                success_count = success
                failure_count = len(failed)

                logging.info(
                    f"Bulk sync: {success_count} succeeded, {failure_count} failed."
                )

                if failure_count:
                    for i, item in enumerate(failed):
                        logging.warning(
                            "Error on document %s: %s", i, item["update"]["error"]
                        )

                return  # Exit after success (even with partial failures)

            except TransportError as sync_error:
                if sync_error.status_code == 429:
                    wait_time = backoff_base**attempt
                    logging.warning(
                        "429 received, retrying in %s seconds...", wait_time
                    )
                    time.sleep(wait_time)
                    attempt += 1
                else:
                    logging.error("Unexpected error during bulk update: %s", sync_error)
                    raise sync_error

        raise Exception("Max retries exceeded for bulk update.")

    def delete_all(self):
        """Delete all indices in Elasticsearch."""
        try:
            LOGGER.info("Deleting all indices...")
            self.client.indices.delete(index="*")
        except Exception as e:
            LOGGER.error("Error deleting all indices: %s", e)
            raise e

    def search_domains(self, body, **kwargs):
        """Search domains index with specified query body."""
        return self.client.search(index=DOMAINS_INDEX, body=body, **kwargs)

    def scroll_domains(self, scroll_id: str, keepalive: str = "2m"):
        """Fetch the next batch of results for a scroll context."""
        return self.client.scroll(scroll_id=scroll_id, params={"scroll": keepalive})

    def clear_scroll_domains(self, scroll_id: str):
        """Clear the scroll context to free resources."""
        return self.client.clear_scroll(body={"scroll_id": [scroll_id]})

    def search_organizations(self, body):
        """Search organizations index with specified query body."""
        return self.client.search(index=ORGANIZATIONS_INDEX, body=body)

    def _bulk_update(self, actions):
        """Update to Elasticsearch."""
        try:
            success_count, response = helpers.bulk(
                self.client, actions, raise_on_error=False
            )

            for idx, item in enumerate(response):
                if "update" in item and item["update"].get("error"):
                    LOGGER.error(
                        "Error indexing document %s: %s", idx, item["update"]["error"]
                    )
                else:
                    LOGGER.info("Successfully indexed document %s: %s", idx, item)

            self.client.indices.refresh(index="domains-5")
        except Exception as e:
            LOGGER.error("Bulk operation error: %s", e)
            raise e
